doc/inlining.md

```diff
@@ -52,74 +78,122 @@ to:
 function main()
   var x = 1
   var call = someExpression
-  osr [call != 'foo] ...
+  assume [call == 'foo] or ...
   call y = 'foo(x)
   print y
 ```
 
 ## With osr
 
-Having the osr in the outer function poses no additional difficulties. Now lets assume the inlinee has safepoints.
-As an example:
+Having the osr in the outer function poses no additional
+difficulties. Now lets assume the inlinee has safepoints.
+
+As an example (we use `rN` for calleR labels, `eN` for calleE label,
+and `iN` for Inlined-code labels):
 
 ```
 function main()
   version base
-    1   var x = 1
-    2   call y = 'foo(x)
-    3   osr sp1 [] (main, base, sp1) [var x = x, var y = y]
-    4   print y
+    r1:   var x = 1
+    r2:   call y = 'foo(x) -> l
+    r3:   print y
+    r4:   stop
 
 function foo(var z)
   version base
-    1   osr sp2 [] (foo, base, sp2) [var z = z]
-    2   return z
+    e1:   assume [...] or (bar, base, checkpoint) [var z = z], cont-frames
+    e2:   return z
 ```
 
-The safepoints of the inlinee need to be extended to recreate the correct number of stack frames.
 Not accounting for the additional stackframe the (incorrect) optimized version looks like this:
 
 ```
 function main()
   version opt
-    1   var x = 1
-    2a  var z_1 = x
-    2b  osr sp2 [] (foo, base, sp2) [var z = z_1]   # wrong
-    2c  var y = z_1
-    2d  drop x_1
-    3   osr sp1 [] (main, base, sp1) [var x = x, var y = y]
-    4   print y
-  ...
-
-function foo(var z)
-  version base
-    1   osr sp2 [] (foo, base, sp2) [var z = z]
-    2   return z
+        var x = 1
+        # prologue
+        var z' = z
+        var y = nil
+    i1: assume [...] or (bar, base, checkpoint) varmap, cont-frames # wrong
```

Comment:
Wait, this specific example had the call inlined, so there is no call instruction anymore and thus no `-> l`. One example above (in the function-specialization case, where no inlining is done) didn't have the return label, but I added it for your sake :-)